File: app/src/main/java/com/raredev/theblocklogics/viewmodel/MainViewModel.java
package com.raredev.theblocklogics.viewmodel;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;
import com.raredev.theblocklogics.R;
import com.raredev.theblocklogics.models.Project;
import com.raredev.theblocklogics.tasks.Task;
import com.raredev.theblocklogics.utils.Constants;
import com.raredev.theblocklogics.utils.loaders.ProjectsLoader;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class MainViewModel extends ViewModel {

  // MainActivity

  private MutableLiveData<Integer> previousFragment = new MutableLiveData<>(-1);
  private MutableLiveData<Integer> selectedFragment = new MutableLiveData<>(-1);

  private MutableLiveData<Integer> navigationSelectedItemId = new MutableLiveData<>(R.id.menu_home);

  public MutableLiveData<Integer> getPreviousFragment() {
    return this.previousFragment;
  }

  public LiveData<Integer> getSelectedFragment() {
    return selectedFragment;
  }

  public void setSelectedFragment(int fragment) {
    var previousFragment = selectedFragment.getValue();

    if (fragment != previousFragment) {
      this.previousFragment.setValue(previousFragment);
      selectedFragment.setValue(fragment);

      if (fragment == Constants.HOME_FRAGMENT) {
        setNavigationSelectedItemId(R.id.menu_home);
      } else if (fragment == Constants.SETTINGS_FRAGMENT) {
        setNavigationSelectedItemId(R.id.menu_settings);
      }
    }
  }

  public void observeSelectedFragment(LifecycleOwner owner, Observer<Integer> observer) {
    selectedFragment.observe(owner, observer);
  }

  public LiveData<Integer> getNavigationSelectedItemId() {
    return this.navigationSelectedItemId;
  }

  public void setNavigationSelectedItemId(int selectedItemId) {
    int previousItemId = this.navigationSelectedItemId.getValue();
    if (previousItemId != selectedItemId) {
      this.navigationSelectedItemId.setValue(selectedItemId);
    }
  }

  public void observeNavigationSelectedItemId(LifecycleOwner owner, Observer<Integer> observer) {
    this.navigationSelectedItemId.observe(owner, observer);
  }

  // HomeFragment

  private MutableLiveData<List<Project>> projects = new MutableLiveData<>(new ArrayList<>());

  public LiveData<List<Project>> getProjects() {
    return this.projects;
  }

  public List<Project> getProjectsList() {
    return projects.getValue();
  }

  public void setPojectsList(List<Project> projects) {
    this.projects.setValue(projects);
  }

  public void loadProjects() {
    if (getSelectedFragment().getValue() != Constants.HOME_FRAGMENT) {
      return;
    }
    new LoadProjectsTask(this).start();
  }

  class LoadProjectsTask extends Task<List<Project>> {

    private WeakReference<MainViewModel> weakReferenceViewModel;

    public LoadProjectsTask(MainViewModel viewModel) {
      weakReferenceViewModel = new WeakReference<>(viewModel);
    }

    @Override
    public List<Project> doInBackground() throws Exception {
      return ProjectsLoader.fetchProjects();
    }

    @Override
    public void onSuccess(List<Project> result) {
      var viewModel = weakReferenceViewModel.get();
      if (viewModel != null) {
        viewModel.setPojectsList(result);
      }
    }
  }
}

File: app/src/main/java/com/raredev/theblocklogics/viewmodel/MainViewModel.kt
package com.raredev.theblocklogics.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.raredev.theblocklogics.R
import com.raredev.theblocklogics.models.Project
import com.raredev.theblocklogics.utils.Constants
import com.raredev.theblocklogics.utils.loaders.ProjectsLoader
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.launch

class MainViewModel: ViewModel() {

  // MainActivity

  private val _previousFragment = MutableLiveData(-1)
  private val _selectedFragment = MutableLiveData(-1)

  val previousFragment: LiveData<Int> = _previousFragment
  val selectedFragment: LiveData<Int> = _selectedFragment

  private val _navSelectedItemId = MutableLiveData(R.id.menu_home)

  val navSelectedItemId: LiveData<Int> = _navSelectedItemId

  fun setFragment(fragment: Int) {
    val previous = _selectedFragment.value

    if (fragment != previous) {
      _previousFragment.value = previous
      _selectedFragment.value = fragment

      val navItemId = when (fragment) {
        Constants.HOME_FRAGMENT -> R.id.menu_home
        Constants.SETTINGS_FRAGMENT -> R.id.menu_settings
        else -> throw IllegalArgumentException("Invalid fragment: $fragment")
      }
      setNavigationSelectedItem(navItemId)
    }
  }

  fun setNavigationSelectedItem(itemId: Int) {
    val previousItemId = _navSelectedItemId.value
    if (itemId != previousItemId) {
      _navSelectedItemId.value = itemId
    }
  }

  // HomeFragment

  private val _projects = MutableLiveData<List<Project>?>(null)

  val projects: LiveData<List<Project>?> = _projects

  fun setProjects(projects: List<Project>) {
    this._projects.value = projects
  }

  fun loadProjects() {
    if (_selectedFragment.value == Constants.HOME_FRAGMENT) {
      CoroutineScope(Dispatchers.IO).launch {
        val projects = ProjectsLoader.fetchProjects()

        withContext(Dispatchers.Main) {
          setProjects(projects)
        }
      }
    }
  }
}